### ISSUE TITLE
Question: Do we need all subtype codecs for the ADT?

### ISSUE THREAD

**Issue by mushtaq** [sirthias/borer#35]
We have an ADT like this, for which we have written the codec only for the base type and it compiles successfully. 

```scala
sealed trait A
case class B(a: Int) extends A
case class C(a: Int) extends A

implicit val a: Codec[A] = deriveCodec[A]
```

If we want to serialize/deserialize the values as (compile-time) type `A`, it works fine. 

The docs suggest that we should write similar codecs for `B` and `C` too, but I guess that is only if we want to use specific compile-time types for ser/de. Is this correct? Is there any performance difference?

Maybe, something to clarify in the docs.

**Comment by sirthias** [comment]
Yes, this effect is indeed unintended and an artifact of borer 0.9.0 still relying on magnolia for codec derivation. magnolia _does_ automatically construct typeclasses for subtypes even if the method calling its macro is not marked `implicit`.

As of the (almost ready) borer release 0.10.0 codec derivation does not rely on magnolia anymore and as such this effect will go away. Then you indeed _will_ have to provide one line per ADT subtype.

So, in 0.9.0 ADT subtype codecs _can_ be automatically derived but this is not recommended for two reasons:

1. Upgrading to the coming 0.10.0 will require adding the missing `deriveCodec` calls.
2. The automatically derived codecs are not cached, which is not a problem in simple cases but can create quite a few headaches in larger, more complicated ADT structures.

**Comment by mushtaq** [comment]
@sirthias thanks for the explanation. But I do not fully understand the second point.

> 2. The automatically derived codecs are not cached, which is not a problem in simple cases but can create quite a few headaches in larger, more complicated ADT structures.

If I explicitly write `implicit val a: Codec[A] = deriveCodec[A]`, I am already caching the codec. So if instead of expecting that subtype codecs exist for delegation, what if borer inlines the subtype codecs also within the `val a`? That way, it will be cached for all the invocation without the boilerplate.

Maybe, instead of changing the default way, we can add something like `Codec.deriveAdt[A]`? Our codebase which has many subtypes for each ADT will immensely benefit from this feature!

**Comment by sirthias** [comment]
The "not cached" problems for example here:

```scala
sealed trait Expr
sealed trait SubExpr extends Expr
case class Addition(left: SubExpr, right: SubExpr) extends Expr
case class Multiplication(left: SubExpr, right: SubExpr) extends Expr
```

If you `deriveCodec[Expr]` then magnolia will re-derive the typeclass for `SubExpr` for every parameter with that type again, which can become a real problem.
Check out this ticket for more info: https://github.com/propensive/magnolia/issues/79

borer 0.10.0 will not fix this but it will _require_ for you to provide the implicit codec for `SubExpr` implicitly, ideally as a cached `val`, which make this problem go away...

**Comment by mushtaq** [comment]
Ah, I see. 

Just to be clear, in a simple case like I mention above, there is no recursion in defining `A`'s subtypes. Will 0.10.0 still require explicit codecs for `B` and `C`?

**Comment by sirthias** [comment]
Yes, for now borer implements "semi-automatic" type class derivation only, which means that we have to write one line per derived type. Fully automatic derivation comes with quite a few problems and intricacies (as the magnolia ticket list shows). Also, error reporting and debugging issues becomes quite a bit more complex and there is a risk of code size explosion that can be quite hard to detect.

Even though magnolia attempts to provide fully automatic derivation we have constrained ourselves to the semi-automatic use case even with magnolia, as we've been bitten quite a few times by issues stemming from the fully automatic derivation that cost us a lot of debugging time...

The "explicit implicits" approach requires a few more lines, yes, but since these are quite uniform they are usually not hard to write. Also, it makes it easy to add customizations to specific sub-types, which tend to creep up over time and create additional issues in the fully-automatic case.
(Because "fully automatic" tends to end up being "fully automatic except for here and here and here".)

**Comment by plokhotnyuk** [comment]
@sirthias could you please give a code sample when "fully automatic" derivation doesn't work for the latest version of borer (w/o magnolia)?

**Comment by sirthias** [comment]
@plokhotnyuk In the example that @mushtaq provided with the original description of this ticket you'll have to write this in order to make it compile:

```scala
implicit val d = deriveCodec[D]
implicit val c = deriveCodec[C]
implicit val b = deriveCodec[B]
implicit val a = deriveCodec[A]
```

The last line alone will not suffice.

**Comment by plokhotnyuk** [comment]
Ok, let me rephrase the question: which issue or contradiction prevents adding of "fully automatic" derivation to the current version of borer that now doesn't depend on magnolia?

As an example, jsoniter-scala uses "fully automatic" approach from the beginning and allows [to customize codecs for any nested type](https://github.com/plokhotnyuk/jsoniter-scala/blob/2f61b1bdc209f88f40286642dd12a898d68d0ff6/jsoniter-scala-macros/src/test/scala/com/github/plokhotnyuk/jsoniter_scala/macros/JsonCodecMakerSpec.scala#L413-L475) with the help of implicits (not only root one)...

**Comment by sirthias** [comment]
I'm not saying that fully automatic derivation isn't possible.
It sure is.
It's just that it's complexities are not something that I wanted to tackle at this point in borer's development. IMHO the semi-automatic approach provides 80% of the value with 20% of the bug surface...

**Comment by mushtaq** [comment]
Some more context to motivate our use case. Currently we are using automatic kryo serialisation via twitter/chill library for Akka remote messages. Given that twitter/chill library is not as actively maintained, we thought of migrating to `jackson-cbor` that now ships with Akka 2.6. But even that requires annotations for ADTs!

So we are thinking of completely migrating to static codecs in `borer` and implement Akka's Serializer interface ourselves. It looks very much feasible, but we have a few ADTs like [this one](https://github.com/tmtsoftware/csw/blob/master/csw-params/shared/src/main/scala/csw/params/commands/CommandIssue.scala). 

If one could write a single line `implicit val a: Codec[CommandResponse] = deriveCodec[CommandResponse]` in such cases (instead of 16), `borer` will indeed be a very attractive option for Akka remote messages too!

**Comment by sirthias** [comment]
ok, thank you for this additional info, @mushtaq 
I have just added https://github.com/sirthias/borer/issues/36 to track progress on this feature request.
